client: reconcile inconsistent RangeLookup results or retry, never fatal

nvb opened the pull request:
Fixes #19147.

RangeLookup scans can span multiple ranges (#18970). Traditionally, in order to
see a fully-consistent snapshot of multiple ranges, a scan needs to operate in a
Txn with a fixed timestamp. Without this, the scan may read results from
different ranges at different times, resulting in an inconsistent view. This is
why `DistSender` returns `OpRequiresTxnError` for consistent scans outside of
Txns that span multiple ranges.

For `RangeLookups`, a consistent but outdated result is just as useless as an
inconsistent result. Because of this, we allow both inconsistent scans and
consistent scans outside of Txns for `RangeLookups`. In doing so, we need to
make sure that we either reconcile any inconsistencies due to races or retry the
scan if this is not possible. This is necessary because inconsistent results are
possible if `RangeLookups` race with splits. This change ensures that we handle
these races correctly.

It also introduces a consistency-elevation scheme so that an INCONSISTENT scan
is elevated to a CONSISTENT scan if the first `RangeLookups` fails. This
improves the chance that the second `RangeLookups` gets consistent results and
ensures that we don't continue sending requests to an out-of-date follower. This
is something we can revisit when we have a proper follower-read implementation.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21276)
<!-- Reviewable:end -->

bdarnell commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/internal/client/range_lookup.go, line 188 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276#-L2IsT405A8xpO56u9o7:-L2IsT405A8xpO56u9o8:b-y52734) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5471791560a222f782828125a2df7b5e00253c8/pkg/internal/client/range_lookup.go#L188)):*
> ```Go
> 	// RangeDescriptors across two different meta2 ranges. Because these meta2
> 	// writes are transactional, performing the entire scan again immediately
> 	// will not run into the same race.
> ```

Why not? Some time passes between the writing of the intents and their resolution. 

---

*[pkg/internal/client/range_lookup.go, line 257 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276#-L2IseZZ5svlzu6w0M05:-L2IseZZ5svlzu6w0M06:bts9cyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5471791560a222f782828125a2df7b5e00253c8/pkg/internal/client/range_lookup.go#L257)):*
> ```Go
> 		// TODO DURING REVIEW: @bdarnell is it ok to upgrade the consistency
> 		// level in all cases? Aren't there times where we need to scan
> 		// inconsistently (e.g. during splits)? Will these ever hit this code
> ```

Yes, there are times that we need to read inconsistently, because we need to see where a range lives to be able to resolve the meta2 intent (after both splits and replica changes). I'm not sure whether they might hit this code path or not, but it's safer to assume that they could. 

Maybe what we need here is to split up INCONSISTENT mode: send it to the lease holder but keep the current behavior with respect to intents. In fact, since range lookups are the only current use of inconsistent reads, maybe we should just redefine INCONSISTENT mode that way (rename it to READ_UNCOMITTED?) and remove the logic to allow inconsistent reads from followers. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/internal/client/range_lookup.go, line 257 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276#-L2IseZZ5svlzu6w0M05:-L2LKsZBbvOJmJ2TZ2CO:b-yul8ib) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5471791560a222f782828125a2df7b5e00253c8/pkg/internal/client/range_lookup.go#L257)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yes, there are times that we need to read inconsistently, because we need to see where a range lives to be able to resolve the meta2 intent (after both splits and replica changes). I'm not sure whether they might hit this code path or not, but it's safer to assume that they could. 

Maybe what we need here is to split up INCONSISTENT mode: send it to the lease holder but keep the current behavior with respect to intents. In fact, since range lookups are the only current use of inconsistent reads, maybe we should just redefine INCONSISTENT mode that way (rename it to READ_UNCOMITTED?) and remove the logic to allow inconsistent reads from followers. 
</blockquote></details>

I like @bdarnell's idea. Reading from the lease holder while scooping up/ignoring intents is what we want and I don't see us using `INCONSISTENT` for anything else in the near future.

What would we lose if we made `RangeLookup` always `CONSISTENT`? At the MVCC level, the command could still special-case intents (and never return them as such, so we'd never fire up the intent resolver machinery). There's probably something I need to page back in.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
> What would we lose if we made RangeLookup always CONSISTENT?

Currently, it's not allowed to send ReturnIntents with a CONSISTENT Scan. If we changed it so that ReturnIntents worked with CONSISTENT (and had the current MVCC-level behavior of ReturnIntents+INCONSISTENT), it would work (modulo migration/mixed-version concerns)

nvanbenschoten commented:
>In fact, since range lookups are the only current use of inconsistent reads, maybe we should just redefine INCONSISTENT mode that way (rename it to READ_UNCOMITTED?) and remove the logic to allow inconsistent reads from followers.

I tried making this change but ran into some issues. Making `INCONSISTENT` reads go the leaseholder was straightforward enough because only a few tests relied on it, but making `INCONSISTENT` reads require a lease (calling through `redirectOnOrAcquireLease`) was much more difficult. Throughout our tests, we peek into the state of follower replicas by sending them inconsistent read requests. If we change the behavior of `INCONSISTENT` reads to require the lease then we lose this ability and break a lot of tests.

I think the better approach is going to be to add a new `READ_UNCOMITTED` read consistency type that goes to the leaseholder and requires a lease like `CONSISTENT` reads but avoids the `CommandQueue` and `TimestampCache` like `INCONSISTENT` reads. This new read consistency type could replace the `ScanRequest.ReturnIntents` option and be used for `RangeLookups` (pending migration/mixed-version concerns, like you said).

nvb committed:
client: use READ_UNCOMMITTED scans in RangeLookup

Now that we have a READ_UNCOMMITTED read consistency level, we
can use this to improve the behavior of RangeLookup scans. This
change switches all DistSender initiated RangeLookup scans over
from INCONSISTENT to READ_UNCOMMITTED as long as the cluster
version allows it. This needs to be behind a cluster version
check because old binaries won't know about READ_UNCOMMITTED.

Release note: None

nvb committed:
client: improve RangeLookup request detection

Release note: None

nvb committed:
storage: add ForceSyncIntentResolution option to StoreTestingKnobs

Release note: None

nvb committed:
client: reconcile inconsistent RangeLookup results or retry, never fatal

Fixes #19147.

RangeLookup scans can span multiple ranges (#18970). Traditionally, in order to
see a fully-consistent snapshot of multiple ranges, a scan needs to operate in a
Txn with a fixed timestamp. Without this, the scan may read results from
different ranges at different times, resulting in an inconsistent view. This is
why `DistSender` returns `OpRequiresTxnError` for consistent scans outside of
Txns that span multiple ranges.

For `RangeLookups`, a consistent but outdated result is just as useless as an
inconsistent result. Because of this, we allow both inconsistent scans and
consistent scans outside of Txns for `RangeLookups`. In doing so, we need to
make sure that we either reconcile any inconsistencies due to races or retry the
scan if this is not possible. This is necessary because inconsistent results are
possible if `RangeLookups` race with splits. This change ensures that we handle
these races correctly.

Release note: None

nvanbenschoten commented:
This change has been rebased on top of https://github.com/cockroachdb/cockroach/pull/22541. RangeLookups now use the new `READ_UNCOMMITTED` read consistency level when possible, and we no longer need to elevate the consistency level in the RangeLookup retry loop. PTAL.

---

Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/internal/client/range_lookup.go, line 188 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276#-L2IsT405A8xpO56u9o7:-L51i4ZCXH7MtxLrNFkk:b-4sdpwd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5471791560a222f782828125a2df7b5e00253c8/pkg/internal/client/range_lookup.go#L188)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why not? Some time passes between the writing of the intents and their resolution. 
</blockquote></details>

These range lookup scans return both committed and uncommitted values, so if intents have been laid down but not resolved, we'll be fine. The race (discussed in more detail in `TestStoreSplitRangeLookupRace`) requires that the split lay down intents and resolve them all between the RangeLookup scan of the first and second meta2 range. If we then rescan the first meta2 range, we're guaranteed by causality to see the result of the split. That's doesn't mean we won't see another race, but the same split will never cause us to see inconsistent results twice.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276#-:-L59do41vLOcEkJgvxBX:bpj0gwg)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 7 files at r2, 1 of 2 files at r4, 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/internal/client/range_lookup.go, line 188 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276#-L2IsT405A8xpO56u9o7:-L5FCCgQJ8L71TrNn6bl:b-53e2v7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5471791560a222f782828125a2df7b5e00253c8/pkg/internal/client/range_lookup.go#L188)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

These range lookup scans return both committed and uncommitted values, so if intents have been laid down but not resolved, we'll be fine. The race (discussed in more detail in `TestStoreSplitRangeLookupRace`) requires that the split lay down intents and resolve them all between the RangeLookup scan of the first and second meta2 range. If we then rescan the first meta2 range, we're guaranteed by causality to see the result of the split. That's doesn't mean we won't see another race, but the same split will never cause us to see inconsistent results twice.
</blockquote></details>

One split won't cause us to see inconsistent results twice, but couldn't there be a second split? (although the code this comment was referring to doesn't seem to be here any more)

---

*[pkg/internal/client/range_lookup.go, line 566 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276#-L5FDITFHT_vPPG-lxqF:-L5FDITFHT_vPPG-lxqG:bquizuv) ([raw file](https://github.com/cockroachdb/cockroach/blob/65aa4fd2d7e0bea491df850db218bfb1afe7a026/pkg/internal/client/range_lookup.go#L566)):*
> ```Go
> var rangeLookupStartKeyBounds = roachpb.Span{
> 	Key:    keys.Meta1Prefix,
> 	EndKey: keys.Meta2KeyMax.Next(),
> ```

Per #9112, I think we'd prefer `Meta2Prefix.PrefixEnd()` to `Meta2KeyMax.Next()`. (although I'm not positive; this is an atypical use of `Span`)

---

*[pkg/internal/client/range_lookup.go, line 570 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276#-L5FDQ2xNAge15CdCscP:-L5FDQ2xNAge15CdCscQ:bme4otl) ([raw file](https://github.com/cockroachdb/cockroach/blob/65aa4fd2d7e0bea491df850db218bfb1afe7a026/pkg/internal/client/range_lookup.go#L570)):*
> ```Go
> var rangeLookupEndKeyBounds = roachpb.Span{
> 	Key:    keys.Meta1Prefix.Next(),
> 	EndKey: keys.SystemPrefix.Next(),
> ```

Ditto: use PrefixEnd instead of next for the end key. 

---

*[pkg/internal/client/range_lookup_test.go, line 124 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276#-L5FE879x_aBAOR6tCse:-L5FE879x_aBAOR6tCsf:b-5pds3) ([raw file](https://github.com/cockroachdb/cockroach/blob/65aa4fd2d7e0bea491df850db218bfb1afe7a026/pkg/internal/client/range_lookup_test.go#L124)):*
> ```Go
> 	})
> 
> 	// This scenerio is similar to the previous one, but assumes that a scan
> ```

s/scenerio/scenario/

---

*[pkg/storage/client_split_test.go, line 2488 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276#-L5FF7nUbQcIko3bEMsJ:-L5FF7nUbQcIko3bEMsK:bdojjre) ([raw file](https://github.com/cockroachdb/cockroach/blob/65aa4fd2d7e0bea491df850db218bfb1afe7a026/pkg/storage/client_split_test.go#L2488)):*
> ```Go
> // multiple meta2 ranges that races with a split and misses all matching
> // descriptors will retry its scan until it succeeds.
> func TestStoreSplitRangeLookupRace(t *testing.T) {
> ```

What is the relationship between this test and the one in internal/client? From the comments they sound redundant, but the implementations look different. Add comments about what each one is testing so we don't think they're redundant in the future (unless they're already redundant now and we can delete one)

---

*[pkg/storage/client_split_test.go, line 2491 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276#-L5FF1pKLDxDPgkNHJox:-L5FF1pKLDxDPgkNHJoy:b-5pds3) ([raw file](https://github.com/cockroachdb/cockroach/blob/65aa4fd2d7e0bea491df850db218bfb1afe7a026/pkg/storage/client_split_test.go#L2491)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 
> 	// The scenerio is modeled after:
> ```

s/scenerio/scenario/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276#-:-L5FFfr99Av6ITWC3AeU:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
TFTR!

---

Review status: 0 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/internal/client/range_lookup.go, line 188 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276#-L2IsT405A8xpO56u9o7:-L5PxTiIK1yrJe0G6ySs:b8u0m0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5471791560a222f782828125a2df7b5e00253c8/pkg/internal/client/range_lookup.go#L188)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

One split won't cause us to see inconsistent results twice, but couldn't there be a second split? (although the code this comment was referring to doesn't seem to be here any more)
</blockquote></details>

Yeah that's definitely true, we could run into a race from a new split on the next iteration. That's why I said we will not run into the same race.

---

*[pkg/internal/client/range_lookup.go, line 566 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276#-L5FDITFHT_vPPG-lxqF:-L5PyRfs5yqNhnAjJunk:b-plg72a) ([raw file](https://github.com/cockroachdb/cockroach/blob/65aa4fd2d7e0bea491df850db218bfb1afe7a026/pkg/internal/client/range_lookup.go#L566)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Per #9112, I think we'd prefer `Meta2Prefix.PrefixEnd()` to `Meta2KeyMax.Next()`. (although I'm not positive; this is an atypical use of `Span`)
</blockquote></details>

`Next()` is used to make these spans end-inclusive for `Span.ContainsKey`. Added a comment.

---

*[pkg/internal/client/range_lookup_test.go, line 124 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276#-L5FE879x_aBAOR6tCse:-L5Q-Yzemb5hhIeVE8SN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/65aa4fd2d7e0bea491df850db218bfb1afe7a026/pkg/internal/client/range_lookup_test.go#L124)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/scenerio/scenario/
</blockquote></details>

Done.

---

*[pkg/storage/client_split_test.go, line 2488 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276#-L5FF7nUbQcIko3bEMsJ:-L5Q-cKTkQc1EPGhQjol:b-2zrhd3) ([raw file](https://github.com/cockroachdb/cockroach/blob/65aa4fd2d7e0bea491df850db218bfb1afe7a026/pkg/storage/client_split_test.go#L2488)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

What is the relationship between this test and the one in internal/client? From the comments they sound redundant, but the implementations look different. Add comments about what each one is testing so we don't think they're redundant in the future (unless they're already redundant now and we can delete one)
</blockquote></details>

This was meant to actually create a series of events that could cause the injected results we saw in the other test. Added a comment here to expand on this difference.

---

*[pkg/storage/client_split_test.go, line 2491 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276#-L5FF1pKLDxDPgkNHJox:-L5Q-_Pda6RK5SU-l1c8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/65aa4fd2d7e0bea491df850db218bfb1afe7a026/pkg/storage/client_split_test.go#L2491)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/scenerio/scenario/
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21276#-:-L5Q3zAfMbRdSnMTh0W5:b-814g8k)*
<!-- Sent from Reviewable.io -->
